cmd/lotus-chainwatch/storage.go

```diff
@@ -318,7 +326,47 @@ create table if not exists miner_sector_events
     
 	constraint miner_sector_events_pk
 		primary key (sector_id, event, miner_id, state_root)
-)
+);
+
+create table if not exists market_deal_proposals
+(
+    deal_id bigint not null,
+    
+    state_root text not null,
+    
+    piece_cid text not null,
+    padded_piece_size bigint not null,
+    unpadded_piece_size bigint not null,
+    verified_deal bool not null,
+    
+    client_id text not null,
+    provider_id text not null,
+    
+    start_epoch bigint not null,
+    end_epoch bigint not null,
+    storage_price_per_epoch text not null,
+    
+    provider_collateral text not null,
+    client_collateral text not null,
+    
+   constraint market_deal_proposal_pk
+ 		primary key (deal_id)
+);
+
+create table if not exists market_deal_states 
+(
+    deal_id bigint not null,
+    
+    sector_start_epoch bigint not null,
+    last_update_epoch bigint not null,
+    slash_epoch bigint not null,
+    
+    state_root text not null,
+    
+	constraint market_deal_states_pk
+		primary key (deal_id,sector_start_epoch,last_update_epoch,slash_epoch)
```

Comment:
nit: Add spaces after commas for readability?

Reply:
We don't really need all of the epochs in the index. I would do primary key as `deal_id, state_root` and if we think there will be a lot of searches on epoch, then we add separate indexes for this.

Reply:
I'm using (abusing) these fields so that only changes to deal state are persisted.

If only the `deal_id` and `state_root` are used as primary keys I'll need to inspect the `DealState` amt at every epoch for changes (using the code [here](https://github.com/filecoin-project/lotus/pull/2322/files#diff-36a0e68e9776295b8ce1875a5e8a1954R250)). Performing this inspection will be slow. 

One compromise here could be creating a separate index on `deal_id, state_root` as you've suggested, what do you think?

Reply:
I prefer the primary key to be as recommended `deal_id, state_root` and create an extra unique constraint across all to achieve the effect you're after.